### PR TITLE
 Add ARM Support for FastPFOR using SIMDe

### PR DESCRIPTION
**Pull request by seb711**
This pull request introduces support for ARM architecture in the FastPFOR project, particularly for building on aarch64 devices. The need for ARM support arose while running FastPFOR on a Graviton Instance. After conducting research and consulting with my mentor at Technical University of Munich (TUM), we decided to integrate SIMDe.

This is a draft pull request, primarily to seek feedback and further insights. The current state of the code might not fully align with the repository's requirements, and due to time constraints, I couldn't perform exhaustive tests (FastPFOR_unittest and "make check" run without errors though). The purpose of this pull request is to share the progress with the community, especially for those interested in building the FastPFOR library on an AARCH64 machine.

### CODE REVIEW

**Comment by lemire** [comment]
Sure, this can be merged, if the tests are passing.

Please see the failing tests.

File: src/simdunalignedbitpacking.cpp [hunk 1]
```diff
@@ -8923,15 +8923,15 @@ static void __SIMD_fastpack16_32(const uint32_t *__restrict__ _in,
   }
 }
 
-static void __SIMD_fastunpack1_32(const __m128i *__restrict__ in,
-                                  uint32_t *__restrict__ _out) {
-  __m128i *out = reinterpret_cast<__m128i *>(_out);
```
Comment (PatrickvL) [inline src/simdunalignedbitpacking.cpp:8928]: Restoring original indentation would avoid unnecessary changes in this pull request, here and elsewhere.

Comment (seb711) [reply src/simdunalignedbitpacking.cpp]: Thanks! Restored the original indentation!

File: src/simdunalignedbitpacking.cpp [hunk 2]
```diff
@@ -8943,7 +8943,73 @@ static void __SIMD_fastunpack1_32(const __m128i *__restrict__ in,
...
+    OutReg1 = _mm_and_si128(_mm_srli_epi32(InReg1, 28), mask);
+    OutReg2 = _mm_and_si128(_mm_srli_epi32(InReg2, 29), mask);
+    OutReg3 = _mm_and_si128(_mm_srli_epi32(InReg1, 30), mask);
+    OutReg4 = _mm_and_si128(_mm_srli_epi32(InReg2, 31), mask);
+    _mm_store_si128(out++, OutReg1);
+    _mm_store_si128(out++, OutReg2);
+    _mm_store_si128(out++, OutReg3);
+    _mm_store_si128(out++, OutReg4);
+#endif
+    }
```
Comment (PatrickvL) [inline src/simdunalignedbitpacking.cpp:9012]: Indentation-only change detected ;)

Comment (PatrickvL) [reply src/simdunalignedbitpacking.cpp]: Code style reviews are one thing. Functional reviews another. I've looked at style, but don't feel comfortable to review functionality.

File: headers/simdgroupsimple.h [hunk 3]
```diff
@@ -99,796 +99,1353 @@ namespace FastPForLib {
...
+#elif defined(__aarch64__)
+        inline static void comprIncompleteBlock(const uint8_t &n, const __m128i *&in,
+                                                  __m128i *&out) {
+            // Since we have to produce exactly one compressed vector anyway, we can
+            // use the highest bit width allowing us to pack all n values.
+            const unsigned b = 32 / n;
+            __m128i comprBlock = _mm_load_si128(in++);
+            for (size_t k = 1; k < n; k++)
+              comprBlock = _mm_or_si128(comprBlock,
+                                        mm_slli_epi32_unrolled(_mm_load_si128(in++), k * b));
```
Comment (lemire) [inline headers/simdgroupsimple.h:178]: Can you provide an assembly dump and/or a before/after benchmark?

It seems that you are putting a lot of trust in the optimizing compiler. This might be warranted, but I think you have to verify.

File: headers/simdgroupsimple.h [hunk 4]
```diff
@@ -99,796 +99,1353 @@ namespace FastPForLib {
...
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 24), mask));
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 25), mask));
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 26), mask));
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 27), mask));
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 28), mask));
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 29), mask));
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 30), mask));
-    _mm_storeu_si128(out++, _mm_and_si128(_mm_srli_epi32(comprBlock, 31), mask));
+      _mm_store_si128(out++,
+                      _mm_and_si128(mm_srli_epi32_unrolled(comprBlock, k * b), mask));
```
Comment (lemire) [inline headers/simdgroupsimple.h:654]: Same question here, you are putting your faith in the optimizing compiler, but you should verify that the perf. in preserved. This can be checked with a before/after benchmark (e.g., on an x64 processor).

File: headers/simdgroupsimple.h (hunk 3 above)
Comment (seb711) [reply headers/simdgroupsimple.h:178]: The reason why i had to unroll this operation was that _mm_slli_epi32's second argument has to be a compile time constant in the SIMDe library, because the corresponding SIMD instruction in NEON wants a compile time constant there. Because i wanted as few effect as possible to the original assembly i switch to the unrolled version just in the arm version where the SIMDe lib is used.

Comment (lemire) [reply headers/simdgroupsimple.h:178]: Ok, the code here is guarded by `#elif defined(__aarch64__)`. I see this. That part is fine, but there is another instance later where it is x64 code that you rewrote in a manner that might stress the compiler.

Comment (lemire) [reply headers/simdgroupsimple.h:178]: I stand corrected. I was wrong. The other case is also guarded by `#elif defined(__aarch64__)` it seems.

File: headers/simdgroupsimple.h (hunk 4 above)
Comment (lemire) [reply headers/simdgroupsimple.h:654]: Ok. Closing my comment. Answered above.

File: headers/simdgroupsimple.h (hunk 3 above)
Comment (lemire) [reply headers/simdgroupsimple.h:178]: Ok. Closing my comment. Answered above.

**Comment by lemire** [comment]
@seb711 I invite you to make this PR as `ready for review`. We should also add ARM tests in CI, I will add another comment.

**Comment by lemire** [comment]
@seb711 As part of this PR, can you add in .github/workflow a file that could be named `vs17-arm-ci.yml` (or anything you like).

It should have content like so...

```yml
name: VS17-ARM-CI

on: [push, pull_request]

jobs:
  ci:
    name: windows-vs17
    runs-on: windows-latest
    strategy:
      fail-fast: false
      matrix:
        include:
          - {gen: Visual Studio 17 2022, arch: ARM64}
    steps:
      - name: checkout
        uses: actions/checkout@v4
      - name: Use cmake
        run: |
          cmake -G "${{matrix.gen}}" -A ${{ matrix.arch }} -DCMAKE_CROSSCOMPILING=1 -B build  &&
          cmake --build build --verbose

```

This will only check that we can build for ARM under Visual Studio. It is kind of a sanity test.

**Comment by lemire** [comment]
@seb711 Please review a PR that I have issued against your own PR :

https://github.com/seb711/FastPFor/pull/1

Note that I am doing a PR against your own PR, which you should merge if you approve. Then we can go back and merge your PR into lemire/FastPFor.

I think your code is sane and I will merge. Good work !!!

**Comment by seb711** [comment]
Sorry for the major delay (had some busy days at uni and thought this PR would require more work ^^); added the workflow file and merged your branch.

**Comment by lemire** [comment]
@seb711 Can you have a quick look? Seems like it does not build for ARM...

**Comment by seb711** [comment]
Hi, so I've only tested it for GCC and ARM and not VS-Compiler and ARM. Therefore I had to add the predefined compiler flags for MSC. Added the VS-ARM Workflow + Tests to the PR, so that the ARM build gets also tested.

**Comment by lemire** [comment]
@seb711 Ok. We can merge. Please confirm that  you are happy with the PR as it stands.

**Comment by seb711** [comment]
@lemire I'm happy with the PR as it stands by now (also worked with it in the last weeks and experienced no issues). I only have one concern: Until there is no Github Runner, that tests ARM runs with GCC, any future change related to ARM builds with GCC cannot be verified with the runners. Arm Runner are announced (https://github.blog/changelog/2023-10-30-accelerate-your-ci-cd-with-arm-based-hosted-runners-in-github-actions/), but are only in private beta by now .

**Comment by lemire** [comment]
I am merging.
